lib/spack/spack/solver/concretize.lp

```diff
@@ -1007,37 +1007,31 @@ error(100, "Compiler {1}@{2} requested for {0} cannot be found. Set install_miss
 % Compiler flags
 %-----------------------------------------------------------------------------
 
-% propagate flags when compiler match
-can_inherit_flags(Package, Dependency, FlagType)
- :- depends_on(Package, Dependency),
-    node_compiler(Package, CompilerID),
-    node_compiler(Dependency, CompilerID),
-    not attr("node_flag_set", Dependency, FlagType, _),
-    compiler_id(CompilerID),
-    flag_type(FlagType).
+attr("node_flag_propagate", Package, FlagType, Flag, Source) :-
+  attr("node_flag_possible_prop", Package, FlagType, Flag, Source),
+  not attr("node_flag_set", Package, FlagType, _).
 
-node_flag_inherited(Dependency, FlagType, Flag)
- :- attr("node_flag_set", Package, FlagType, Flag), can_inherit_flags(Package, Dependency, FlagType),
-    attr("node_flag_propagate", Package, FlagType).
-% Ensure propagation
-:- node_flag_inherited(Package, FlagType, Flag),
-    can_inherit_flags(Package, Dependency, FlagType),
-    attr("node_flag_propagate", Package, FlagType).
+attr("node_flag_possible_prop", Package, FlagType, Flag, Source) :-
+  node_compiler(Source, CompilerID),
+  node_compiler(Package, CompilerID),
+  depends_on(Parent, Package),
+  attr("node_flag_possible_prop", Parent, _, _, Source),
+  attr("node_flag", Source, FlagType, Flag),
```

Comment:
This has the same issue as the other PR for variants. Once you have a fact like:
```
attr("node_flag_possible_prop", "foo", "cflags", "some flags", "foo")
```
this rule will propagate every flag set on `foo` regardless it's `cflags` or not.